### ISSUE TITLE
regards getMockInstance removed from 1.42

### ISSUE THREAD

**Issue by xeric** [jmockit/jmockit1#563]
Please provide the following information:

* Version of JMockit that was used: 
1.37

* Description of the problem:
getMockInstance deprecated from 1.37 and removed from 1.42

* Check that:

 - [√] The problem can be reproduced by JMockit project members.
  For that, include an example test (perhaps accompanied by a Maven/Gradle build script) which
  can be executed without changes and reproduces the failure.

 - [√] The issue does not fall outside the scope of the project (for example, attempting to use JMockit
  APIs from Groovy or Scala code, or with an Android runtime).

 - [√] The JDK where the problem occurs is a final release, not a development build.

Not sure if getMockInstance had anything complex to implement in JMockit, until version 1.42, it works well, and it resolved some dedicate mockup cases:
1. When the mock target is a final class, and has protected constructor, it is very hard for me to new a instance for such class, MockUp<T>.getMockInstance resolved this limitation easy and perfect, I can get a mocked instance and replace some implementation of class.
2. when the mock target is an interface has many methods to implement, after getMockInstance removed, seems I have to write a simple class implemented all methods with empty or default return value, if I have getMockInstance, I can just get a minimal instance with 1-2 mocked methods.

Any better workaround for above cases?

**Comment by rliesenfeld** [comment]
1. But why would you need to create an instance of such a class? I imagine it has some static factory method; can't that be used? Anyway, `MockUp#getMockInstance()` was never meant for this.
2. Yes, create a class implementing all interface methods (or use an existing implementation class). This is easily done with any Java IDE (including the return of `0`, `false`, `null`, etc.); it's what I do myself.

In summary, `getMockInstance()` was removed because there were no known legitimate use cases. And your examples only confirm the suspicion that it only got used in questionable cases.

**Comment by xeric** [comment]
Thanks for you reply, maybe my description is a little abstract, let me show an example:
In order to make it easy to explain, I write a simple code
FinalLog is test target
`
public class FinalLog {
  protected static final Logger logger = LogManager.getLogManager().getLogger("a");

  public void logSomething() {
    logger.log(Level.ALL, "something");
  }

  public void anotherLog() {
    try {
      logger.isLoggable(Level.ALL);
    } catch (Exception e) {
      logger.log(Level.ALL, "another thing", e);
    }
  }
}
`
And here is my test:
`
public class FinalLogTest1 {

  @BeforeClass
  public void buildTrueLogger() {
    /*final Logger logger = new MockUp<Logger>(){

    }.getMockInstance();*/

    final Logger logger = new FakeLogger("fakeLoggerName", "fakeBundle");

    new MockUp<LogManager>() {
      @Mock
      Logger getLogger(String name) {
        return logger;
      }
    };

    System.out.println(FinalLog.class);
  }

  @Test
  public void testLogSomething() {
    FinalLog log = new FinalLog();
    log.logSomething();

    //verify log function called
  }

  private final class FakeLogger extends Logger {
    FakeLogger(String name, String resourceBundleName) {
      super(name, resourceBundleName);
    }
  }
}
`

Actually, here is the problem:
In our actual code, Logger is a final class and only created by an ManagerWrapper, it has protected constructor, so can't be initialed by new keyword.
Even I can remove final modify for Logger class, and I have to write a FakeLogger class to extend it.
And in test code, I have to initial this Logger instance with parameter, unfortunately, intial this Logger is not so easy, it requires some dedicate input as parameter to initial it, not some free style text.

Before 1.37, it is quite easy for me: call the code block in commented code - get a fake instance for this Logger and just mock log method with powerful support of jmockit invocation to validate function call.

Not quite sure if any better way to test it, or this is just misleading case for getMockInstance?

**Comment by rliesenfeld** [comment]
Thanks. I played with it a little, and got the following alternative tests (all work with JMockit 1.43):
```
import java.util.*;
import java.util.logging.*;
import org.junit.*;
import static org.junit.Assert.*;
import mockit.*;

public final class FinalLogTest {
   @Tested FinalLog log;

   @Test
   public void logSomething_usingMockUpInABadWay() {
      final List<String> loggedMessages = new ArrayList<>();
      new MockUp<Logger>() {
         @Mock
         void log(Level level, String msg) {
            assertSame(Level.ALL, level);
            loggedMessages.add(msg);
         }
      };

      log.logSomething();

      assertEquals(1, loggedMessages.size());
      assertFalse(loggedMessages.get(0).isEmpty());
   }

   @Test
   public void logSomething_usingMockedInASensibleWay(@Mocked final Logger anyLogger) {
      log.logSomething();

      // This is fine, but is testing calls to a logger really necessary?
      new Verifications() {{ anyLogger.log(Level.ALL, anyString); }};
   }

   @Test
   public void aBetterWayToTestLoggingIfReallyNeeded_noMockingOrFaking() {
      // Retrieve and configure the logger to be tested so its output can be verified:
      Logger theLogger = Logger.getLogger("a");
      theLogger.setLevel(Level.ALL);
      final List<String> loggedMessages = new ArrayList<>();
      theLogger.addHandler(new StreamHandler() {
         @Override
         public synchronized void publish(LogRecord record) {
            assertSame(Level.ALL, record.getLevel());
            loggedMessages.add(record.getMessage());
         }
      });

      log.logSomething();

      assertEquals(1, loggedMessages.size());
      assertFalse(loggedMessages.get(0).isEmpty());
   }
}
```

**Comment by xeric** [comment]
I got NPE when run your test on TestNG 6.14.3 + JMockit 1.43
Did not try to run it on JUnit, I think there're some special initialization on TestNG which has static field initialization for logger instance in FinalLog, the value is null, and actually that's why I was trying to mock logger in @BeforeClass.
```
java.lang.NullPointerException: while trying to invoke the method java.util.logging.Logger.log(java.util.logging.Level, java.lang.String) of a null object loaded from static field com.sap.test.jmockit.sample.replacefinal.FinalLog.logger

	at com.sap.test.jmockit.sample.replacefinal.FinalLog.logSomething(FinalLog.java:17)
	at com.sap.test.jmockit.sample.replacefinal.FinalLogTest1.aBetterWayToTestLoggingIfReallyNeeded_noMockingOrFaking(FinalLogTest1.java:64)
```

BTW, if I can write MockUp in instance test level, I have no necessary to create so complex workaround, I just need to prepare a @Mocked Logger logger and put it in return field in LogManager MockUp.
```
public class FinalLogTest1 {
  
  @Mocked
  private Logger fakeLogger;
  
  @Test
  public void testLogSomething() {
    new MockUp<LogManager>() {
      @Mock
      Logger getLogger(String name) {
        return fakeLogger;
      }
    };
    
    FinalLog log = new FinalLog();
    log.logSomething();

    //verify log function called
  }
}
```
But as I debugged, when running into testLogSomething method, the Logger instance in FinalLog has been initialized and can't be replaced with mocked one anymore.

**Comment by rliesenfeld** [comment]
My bad, I forgot to copy & paste the fixed version of the `FinalLog` class:
```
public class FinalLog {
   // Gets the logger in the usual way (not with LogManager):
   protected static final Logger logger = Logger.getLogger("a");

   public void logSomething() {
      logger.log(Level.ALL, "something");
   }

   public void anotherLog() {
      try {
         logger.isLoggable(Level.ALL);
      }
      catch (Exception e) {
         logger.log(Level.ALL, "another thing", e);
      }
   }
}
```

**Comment by xeric** [comment]
Yes, that's the issue, actually, we have so many places that get instance from a XXXManager, and can't directly get it from Class itself, and to change the original source code one by one is not making sense in our huge project. That's why I think the old method getMockInstance is quite useful here.
